### DISCUSSION TITLE
JsonElement (or ExpandoObject) as script input value?

### DISCUSSION THREAD

**Discussion by BhaaLseN** [sebastienros/jint#1766] [Q&A] [answered]
Hey all!

I got some already deserialized json as `System.Text.Json.JsonElement` (and can turn it into an `ExpandoObject` if need-be; it is not typed nor can I write proper CLR types in advance) that I'd like to pass into a script to be processed (basically `Engine.SetValue("input", jsonElement)`, let the script do its work, then grab the completion value).

Obviously, passing either directly into `SetValue` doesn't _really_ do what I expected (or wanted) it to do; thats why I'm here.

I'm sure I could roll my own solution based on various [`Newtonsoft.Json` wrappers](https://github.com/npenin/jint-newtonsoft.json/blob/master/Jint.NewtonsoftWrapper/NTSObjectInstance.cs) that have been floating around for a while, but all of them are for Jint 2.x and desperately need an upgrade to work with Jint 3.x. Besides the fact that it is code that needs to be maintained and potentially updated when Jint releases new versions, so I'd always be behind and chasing changes.

Doesn't hurt to ask I suppose, so: Does Jint have anything built-in that can help me out here (short of serializing the json back into a string and letting Jint parse it again)?

**Comment by lahma** [comment, marked as answer]
If you are just looking for interop capabilities there's some test coverage and configuration for STJ [here](https://github.com/sebastienros/jint/blob/main/Jint.Tests/Runtime/InteropTests.SystemTextJson.cs).

**Reply by BhaaLseN** [reply]
Ooooh, I looked through the tests and completely missed those. Thanks!

It looks like I can take a shortcut and do the following:
```c#
var input = new ObjectWrapper(engine, jsonElement.Deserialize<JsonNode>());
engine.SetValue("input", input);
```

Is there any downside to this; and/or reason why I should use the `Options` approach instead?
(No arrays in there, so I _think_ it should be ok for the inputs I'm working with - not that it would matter, since `SetPrototypeOf` is `internal` and I can't call it anyways)

**Reply by lahma** [reply]
Please do experiment, there probably is no one right answer.

**Reply by lahma** [reply]
And for prototype the Prototype property setter should work.

**Reply by BhaaLseN** [reply]
Alright, I'll see where this takes me.

Do you want me to open an issue about the `internal` usage from tests? Depending on whether it should double as reference code or not, it might be worthwhile revisiting.

**Reply by BhaaLseN** [reply]
Yep, looks like this works.
```diff
             // make JsonArray behave like JS array
             options.Interop.WrapObjectHandler = static (e, target, type) =>
             {
                 var wrapped = new ObjectWrapper(e, target);
                 if (target is JsonArray)
                 {
-                    wrapped.SetPrototypeOf(e.Realm.Intrinsics.Array.PrototypeObject);
+                    wrapped.Prototype = e.Intrinsics.Array.PrototypeObject;
                 }
                 return wrapped;
             };
```

Thanks again!

**Reply by lahma** [reply]
A PR to improve is always better than an issue. Public API test project doesn't have access to internals.
